--- contrib/python/iniconfig/iniconfig/__init__.py	(index)
+++ contrib/python/iniconfig/iniconfig/__init__.py	(working tree)
@@ -1,6 +1,7 @@
 """ brain-dead simple parser for ini-style files.
 (C) Ronny Pfannschmidt, Holger Krekel -- MIT licensed
 """
+import os
 __all__ = ['IniConfig', 'ParseError']
 
 COMMENTCHARS = "#;"
@@ -49,7 +50,15 @@ class IniConfig(object):
     def __init__(self, path, data=None):
         self.path = str(path)  # convenience
         if data is None:
-            f = open(self.path)
+            if os.path.basename(self.path).startswith('pkg:'):
+                import io, pkgutil
+
+                basename = os.path.basename(self.path)
+                _, package, resource = basename.split(':')
+                content = pkgutil.get_data(package, resource)
+                f = io.StringIO(content.decode('utf-8'))
+            else:
+                f = open(self.path)
             try:
                 tokens = self._parse(iter(f))
             finally:
