--- contrib/python/more-itertools/py2/more_itertools/tests/test_more.py	(index)
+++ contrib/python/more-itertools/py2/more_itertools/tests/test_more.py	(working tree)
@@ -122,13 +122,13 @@ class IterOnlyRange:
     raise an ``AttributeError`` rather than ``TypeError`` in Python 2.
 
     >>> r = IterOnlyRange(5)
-    >>> r[0]
+    >>> r[0]  # doctest: +SKIP
     AttributeError: IterOnlyRange instance has no attribute '__getitem__'
 
     Note: In Python 3, ``TypeError`` will be raised because ``object`` is
     inherited implicitly by default.
 
-    >>> r[0]
+    >>> r[0]  # doctest: +SKIP
     TypeError: 'IterOnlyRange' object does not support indexing
     """
     def __init__(self, n):
