--- contrib/python/traitlets/py3/traitlets/config/tests/test_configurable.py	(index)
+++ contrib/python/traitlets/py3/traitlets/config/tests/test_configurable.py	(working tree)
@@ -33,1 +33,1 @@ from traitlets.traitlets import (
-from ...tests._warnings import expected_warnings
+from traitlets.tests._warnings import expected_warnings
--- contrib/python/traitlets/py3/traitlets/config/tests/test_application.py	(index)
+++ contrib/python/traitlets/py3/traitlets/config/tests/test_application.py	(working tree)
@@ -635,2 +635,3 @@ class TestApplication(TestCase):
 
+@mark.skip
 def test_cli_multi_scalar(caplog):
@@ -655,1 +656,1 @@ def test_cli_multi_scalar(caplog):
-        "sub1": ("traitlets.config.tests.test_application.Sub1", "import string"),
+        "sub1": ("__tests__.config.tests.test_application.Sub1", "import string"),
--- contrib/python/traitlets/py3/traitlets/tests/test_traitlets.py	(index)
+++ contrib/python/traitlets/py3/traitlets/tests/test_traitlets.py	(working tree)
@@ -63,1 +63,1 @@ from unittest import TestCase
-from ._warnings import expected_warnings
+from traitlets.tests._warnings import expected_warnings
--- contrib/python/traitlets/py3/traitlets/utils/tests/test_bunch.py	(index)
+++ contrib/python/traitlets/py3/traitlets/utils/tests/test_bunch.py	(working tree)
@@ -1,1 +1,1 @@
-from ..bunch import Bunch
+from traitlets.utils.bunch import Bunch
--- contrib/python/traitlets/py3/traitlets/utils/tests/test_decorators.py	(index)
+++ contrib/python/traitlets/py3/traitlets/utils/tests/test_decorators.py	(working tree)
@@ -4,2 +4,2 @@ from unittest import TestCase
-from ...traitlets import HasTraits, Int, Unicode
+from traitlets.traitlets import HasTraits, Int, Unicode
-from ..decorators import signature_has_traits
+from traitlets.utils.decorators import signature_has_traits
--- contrib/python/traitlets/py3/traitlets/utils/tests/test_importstring.py	(index)
+++ contrib/python/traitlets/py3/traitlets/utils/tests/test_importstring.py	(working tree)
@@ -11,1 +11,1 @@
-from ..importstring import import_item
+from traitlets.utils.importstring import import_item
--- contrib/python/traitlets/py3/traitlets/tests/utils.py	(index)
+++ contrib/python/traitlets/py3/traitlets/tests/utils.py	(working tree)
@@ -1,10 +1,13 @@
 import sys
 from subprocess import PIPE, Popen
+import os
 
 
 def get_output_error_code(cmd):
     """Get stdout, stderr, and exit code from running a command"""
-    p = Popen(cmd, stdout=PIPE, stderr=PIPE)
+    env = os.environ.copy()
+    env["Y_PYTHON_ENTRY_POINT"] = ":main"
+    p = Popen(cmd, stdout=PIPE, stderr=PIPE, env=env)
     out, err = p.communicate()
     out = out.decode("utf8", "replace")  # type:ignore
     err = err.decode("utf8", "replace")  # type:ignore
