--- contrib/python/pytest/_pytest/mark/structures.py	(index)
+++ contrib/python/pytest/_pytest/mark/structures.py	(working tree)
@@ -440,7 +440,7 @@ class NodeKeywords(MappingMixin):
         return "<NodeKeywords for node %s>" % (self.node,)
 
 
-@attr.s(cmp=False, hash=False)
+@attr.s(eq=False, order=False, hash=False)
 class NodeMarkers(object):
     """
     internal structure for storing marks belonging to a node
--- contrib/python/pytest/_pytest/tmpdir.py	(index)
+++ contrib/python/pytest/_pytest/tmpdir.py	(working tree)
@@ -31,7 +31,7 @@ class TempPathFactory(object):
         # using os.path.abspath() to get absolute path instead of resolve() as it
         # does not work the same in all platforms (see #4427)
         # Path.absolute() exists, but it is not public (see https://bugs.python.org/issue25012)
-        convert=attr.converters.optional(
+        converter=attr.converters.optional(
             lambda p: Path(os.path.abspath(six.text_type(p)))
         )
     )
