--- contrib/python/pytest/_pytest/compat.py	(index)
+++ contrib/python/pytest/_pytest/compat.py	(working tree)
@@ -223,7 +223,7 @@ if _PY3:
         if isinstance(val, bytes):
             return _bytes_to_ascii(val)
         else:
-            return val.encode("unicode_escape").decode("ascii")
+            return val
 
 
 else:
@@ -241,11 +241,11 @@ else:
         """
         if isinstance(val, bytes):
             try:
-                return val.encode("ascii")
+                return val.encode("utf-8")
             except UnicodeDecodeError:
-                return val.encode("string-escape")
+                return val.decode("utf-8", "ignore").encode("utf-8", "replace")
         else:
-            return val.encode("unicode-escape")
+            return val.encode("utf-8", "replace")
 
 
 class _PytestWrapper(object):
