--- contrib/deprecated/python/backports.shutil-get-terminal-size/backports/shutil_get_terminal_size/get_terminal_size.py	(index)
+++ contrib/deprecated/python/backports.shutil-get-terminal-size/backports/shutil_get_terminal_size/get_terminal_size.py	(working tree)
@@ -41,7 +41,7 @@ try:
 
         return terminal_size(columns, lines)
 
-except ImportError:
+except (ImportError, OSError):
     import fcntl
     import termios
 
