--- contrib/python/ipython/py3/.dist-info/METADATA	(index)
+++ contrib/python/ipython/py3/.dist-info/METADATA	(working tree)
@@ -28,7 +28,7 @@ Classifier: Programming Language :: Python :: 3 :: Only
 License-File: LICENSE
 Requires-Dist: backcall
 Requires-Dist: decorator
-Requires-Dist: jedi (>=0.16)
+Requires-Dist: jedi (>=0.13)
 Requires-Dist: matplotlib-inline
 Requires-Dist: pickleshare
 Requires-Dist: prompt-toolkit (!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0)
--- contrib/python/ipython/py3/ya.make	(index)
+++ contrib/python/ipython/py3/ya.make	(working tree)
@@ -14,14 +14,29 @@ PEERDIR(
     contrib/python/decorator
     contrib/python/jedi
     contrib/python/matplotlib-inline
-    contrib/python/pexpect
     contrib/python/pickleshare
     contrib/python/prompt-toolkit
     contrib/python/setuptools
     contrib/python/stack-data
     contrib/python/traitlets
 )
 
+IF (OS_WINDOWS)
+    PEERDIR(
+        contrib/python/colorama
+    )
+ELSE()
+    PEERDIR(
+        contrib/python/pexpect
+    )
+ENDIF()
+
+IF (OS_DARWIN)
+    PEERDIR(
+        contrib/python/appnope
+    )
+ENDIF()
+
 NO_LINT()
 
 NO_CHECK_IMPORTS(
