--- contrib/python/ipython/py3/IPython/core/ultratb.py	(index)
+++ contrib/python/ipython/py3/IPython/core/ultratb.py	(working tree)
@@ -239,7 +239,7 @@ class TBTools(colorable.Colorable):
         self.debugger_cls = debugger_cls or debugger.Pdb
 
         if call_pdb:
-            self.pdb = debugger_cls()
+            self.pdb = self.debugger_cls()
         else:
             self.pdb = None
 
