--- contrib/python/pytest/py2/_pytest/assertion/rewrite.py	(index)
+++ contrib/python/pytest/py2/_pytest/assertion/rewrite.py	(working tree)
@@ -4,6 +4,9 @@ from __future__ import absolute_import
 from __future__ import division
 from __future__ import print_function
 
+import warnings
+warnings.filterwarnings("ignore", category=DeprecationWarning, module="_pytest.assertion.rewrite")
+
 import ast
 import errno
 import imp
