--- contrib/python/pytest/py2/_pytest/compat.py	(index)
+++ contrib/python/pytest/py2/_pytest/compat.py	(working tree)
@@ -244,7 +244,7 @@ if _PY3:
             ret = _bytes_to_ascii(val)
         else:
             ret = val
-        return _translate_non_printable(ret)
+        return ret
 
 
 else:
@@ -267,7 +267,7 @@ else:
                 ret = val.decode("utf-8", "ignore")
         else:
             ret = val.encode("utf-8", "replace").decode("utf-8")
-        return _translate_non_printable(ret)
+        return ret
 
 
 class _PytestWrapper(object):
