commit 46b46cea857997531a7b9b4de32556be50bd5aea (HEAD -> pytest-backport-pr-6607)
author: danila-eremin
date: 2022-01-22T16:55:07+03:00

    backport py2

--- contrib/python/pytest/py2/_pytest/python.py	(b87553999455f0084413c5b59139aa9ccb5e3f52)
+++ contrib/python/pytest/py2/_pytest/python.py	(46b46cea857997531a7b9b4de32556be50bd5aea)
@@ -896,7 +896,7 @@ class CallSpec2(object):
 
     @property
     def id(self):
-        return "-".join(map(safe_str, filter(None, self._idlist)))
+        return "-".join(map(safe_str, self._idlist))
 
     def setmulti2(self, valtypes, argnames, valset, id, marks, scopenum, param_index):
         for arg, val in zip(argnames, valset):
