--- contrib/python/tornado/tornado-6/tornado/curl_httpclient.py	(index)
+++ contrib/python/tornado/tornado-6/tornado/curl_httpclient.py	(working tree)
@@ -418,7 +418,17 @@ class CurlAsyncHTTPClient(AsyncHTTPClient):
             curl.setopt(pycurl.SSL_VERIFYPEER, 0)
             curl.setopt(pycurl.SSL_VERIFYHOST, 0)
         if request.ca_certs is not None:
-            curl.setopt(pycurl.CAINFO, request.ca_certs)
+            cafile, capath, cadata = None, None, None
+            if callable(request.ca_certs):
+                cafile, capath, cadata = request.ca_certs()
+            else:
+                cafile = request.ca_certs
+            if cafile is not None:
+                curl.setopt(pycurl.CAINFO, cafile)
+            if capath is not None:
+                curl.setopt(pycurl.CAPATH, capath)
+            if cadata is not None:
+                curl.set_ca_certs(cadata)
         else:
             # There is no way to restore pycurl.CAINFO to its default value
             # (Using unsetopt makes it reject all certificates).
