--- a/include/uchar/uchar.h	(index)
+++ b/include/uchar/uchar.h	(working tree)
@@ -10,11 +10,12 @@ typedef unsigned short char16_t;
 typedef unsigned char32_t;
 #endif
 
-#define __NEED_mbstate_t
-#define __NEED_size_t
-
 #include <features.h>
-#include <bits/alltypes.h>
+
+#define __need_size_t
+#include <stddef.h>
+
+#include <wchar.h>
 
 size_t c16rtomb(char *__restrict, char16_t, mbstate_t *__restrict);
 size_t mbrtoc16(char16_t *__restrict, const char *__restrict, size_t, mbstate_t *__restrict);
