--- contrib/tools/cython/Cython/Utility/Builtins.c	(616d4f88939402d37616bd144d4a45d21fb85ad3)
+++ contrib/tools/cython/Cython/Utility/Builtins.c	(cf9802e6dcf01199298b0d46c1e3e221b2bacafa)
@@ -305,6 +305,11 @@ static PyObject *__Pyx_PyLong_AbsNeg(PyObject *n) {
 
 //////////////////// divmod_int.proto //////////////////
 
+#ifdef _WIN32
+__declspec(selectany)
+#else
+__attribute__((weak))
+#endif
 const {{RETURN_TYPE}} __Pyx_divmod_ERROR_VALUE_{{CFUNC_SUFFIX}} = {-1, -1};
 
 static CYTHON_INLINE {{RETURN_TYPE}} __Pyx_divmod_{{CFUNC_SUFFIX}}({{TYPE}} a, {{TYPE}} b); /*proto*/
