--- contrib/tools/cython/Cython/Compiler/TypeSlots.py	(3fd698c83ff9f2a6d3e547d212fa364ce1cc0385)
+++ contrib/tools/cython/Cython/Compiler/TypeSlots.py	(b4949f970eee8c854300ccfb54870046111161d3)
@@ -905,6 +905,7 @@ slot_table = (
     EmptySlot("tp_finalize", ifdef="PY_VERSION_HEX >= 0x030400a1"),
     EmptySlot("tp_vectorcall", ifdef="PY_VERSION_HEX >= 0x030800b1 && (!CYTHON_COMPILING_IN_PYPY || PYPY_VERSION_NUM >= 0x07030800)"),
     EmptySlot("tp_print", ifdef="PY_VERSION_HEX >= 0x030800b4 && PY_VERSION_HEX < 0x03090000"),
+    EmptySlot("tp_watched", ifdef="PY_VERSION_HEX >= 0x030C0000"),
     # PyPy specific extension - only here to avoid C compiler warnings.
     EmptySlot("tp_pypy_flags", ifdef="CYTHON_COMPILING_IN_PYPY && PY_VERSION_HEX >= 0x03090000 && PY_VERSION_HEX < 0x030a0000"),
 )
