--- contrib/python/ipython/py2/ya.make	(index)
+++ contrib/python/ipython/py2/ya.make	(working tree)
@@ -16,1 +15,0 @@ PEERDIR(
-    contrib/python/pexpect
@@ -21,6 +18,23 @@ PEERDIR(
     contrib/python/traitlets
 )
 
+IF (OS_WINDOWS)
+    PEERDIR(
+        contrib/python/colorama
+        contrib/deprecated/python/win-unicode-console
+    )
+ELSE ()
+    PEERDIR(
+        contrib/python/pexpect
+    )
+ENDIF ()
+
+IF (OS_DARWIN)
+    PEERDIR(
+        contrib/python/appnope
+    )
+ENDIF ()
+
 NO_LINT()
 
 NO_CHECK_IMPORTS(
