--- contrib/python/joblib/joblib/externals/loky/backend/fork_exec.py	(index)
+++ contrib/python/joblib/joblib/externals/loky/backend/fork_exec.py	(working tree)
@@ -39,6 +39,7 @@ def fork_exec(cmd, keep_fds, env=None):
     env = {} if env is None else env
     child_env = os.environ.copy()
     child_env.update(env)
+    child_env['Y_PYTHON_ENTRY_POINT'] = ':main'
 
     pid = os.fork()
     if pid == 0:  # pragma: no cover
