--- contrib/python/joblib/joblib/externals/loky/backend/fork_exec.py	(index)
+++ contrib/python/joblib/joblib/externals/loky/backend/fork_exec.py	(working tree)
@@ -33,6 +33,7 @@ def fork_exec(cmd, keep_fds, env=None):
     # copy the environment variables to set in the child process
     env = env or {}
     child_env = {**os.environ, **env}
+    child_env['Y_PYTHON_ENTRY_POINT'] = ':main'
 
     pid = os.fork()
     if pid == 0:  # pragma: no cover
