--- contrib/python/more-itertools/py3/tests/test_more.py	(index)
+++ contrib/python/more-itertools/py3/tests/test_more.py	(working tree)
@@ -177,13 +177,13 @@ class IterOnlyRange:
     """User-defined iterable class which only support __iter__.

     >>> r = IterOnlyRange(5)
-    >>> r[0]
+    >>> r[0]  # doctest: +SKIP
     AttributeError: IterOnlyRange instance has no attribute '__getitem__'

     Note: In Python 3, ``TypeError`` will be raised because ``object`` is
     inherited implicitly by default.

-    >>> r[0]
+    >>> r[0]  # doctest: +SKIP
     TypeError: 'IterOnlyRange' object does not support indexing
     """
