--- contrib/python/parso/py3/tests/test_python_errors.py	(index)
+++ contrib/python/parso/py3/tests/test_python_errors.py	(working tree)
@@ -28,7 +28,7 @@ def assert_comparison(code, error_code, positions):
     errors = [(error.start_pos, error.code) for error in _get_error_list(code)]
     assert [(pos, error_code) for pos in positions] == errors
 
-
+@pytest.mark.skipif(sys.version_info >= (3, 10), reason="parso don't support Python 3.10 yet")
 @pytest.mark.parametrize('code', FAILING_EXAMPLES)
 def test_python_exception_matches(code):
     wanted, line_nr = _get_actual_exception(code)
