--- contrib/python/scikit-learn/py3/sklearn/svm/src/liblinear/linear.cpp	(index)
+++ contrib/python/scikit-learn/py3/sklearn/svm/src/liblinear/linear.cpp	(working tree)
@@ -41,6 +41,8 @@
 #include "tron.h"
 #include <climits>
 #include <random>
+#define mt_rand linear_mt_rand
+#define set_seed linear_set_seed
 #include "../newrand/newrand.h"
 
 typedef signed char schar;
--- contrib/python/scikit-learn/py3/sklearn/svm/src/liblinear/linear.h	(index)
+++ contrib/python/scikit-learn/py3/sklearn/svm/src/liblinear/linear.h	(working tree)
@@ -49,8 +49,6 @@ struct model
 	int *n_iter;    /* no. of iterations of each class */
 };
 
-void set_seed(unsigned seed);
-
 struct model* train(const struct problem *prob, const struct parameter *param, BlasFunctions *blas_functions);
 void cross_validation(const struct problem *prob, const struct parameter *param, int nr_fold, double *target);
 
--- contrib/python/scikit-learn/py3/sklearn/svm/src/libsvm/svm.cpp	(index)
+++ contrib/python/scikit-learn/py3/sklearn/svm/src/libsvm/svm.cpp	(working tree)
@@ -68,6 +68,8 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <random>
 #include "svm.h"
 #include "_svm_cython_blas_helpers.h"
+#define mt_rand svm_mt_rand
+#define set_seed svm_set_seed
 #include "../newrand/newrand.h"
 
 
