--- contrib/deprecated/python/scandir/_scandir.c	(index)
+++ contrib/deprecated/python/scandir/_scandir.c	(working tree)
@@ -18,6 +18,7 @@ comment):
 
 #ifdef MS_WINDOWS
 #include <windows.h>
+#include <winioctl.h>
 #include "winreparse.h"
 #else
 #include <dirent.h>
@@ -660,7 +661,7 @@ _pystat_fromstructstat(STRUCT_STAT *st)
     return v;
 }
 
-char *PyStructSequence_UnnamedField = "unnamed field";
+//char *PyStructSequence_UnnamedField = "unnamed field";
 
 PyDoc_STRVAR(stat_result__doc__,
 "stat_result: Result from stat, fstat, or lstat.\n\n\
