--- contrib/python/numpy/py3/numpy/core/src/multiarray/compiled_base.c	(index)
+++ contrib/python/numpy/py3/numpy/core/src/multiarray/compiled_base.c	(working tree)
@@ -14,7 +14,7 @@
 #include "alloc.h"
 #include "ctors.h"
 #include "common.h"
-#include "simd/simd.h"
+#include "contrib/python/numpy/py3/numpy/core/src/common/simd/simd.h"
 
 #include <string.h>
 
--- contrib/python/numpy/py3/numpy/core/src/multiarray/einsum_sumprod.c	(index)
+++ contrib/python/numpy/py3/numpy/core/src/multiarray/einsum_sumprod.c	(working tree)
@@ -27,7 +27,7 @@
 
 #include "einsum_sumprod.h"
 #include "einsum_debug.h"
-#include "simd/simd.h"
+#include "contrib/python/numpy/py3/numpy/core/src/common/simd/simd.h"
 #include "common.h"
 
 // ARM/Neon don't have instructions for aligned memory access
--- contrib/python/numpy/py3/numpy/core/src/multiarray/item_selection.c	(index)
+++ contrib/python/numpy/py3/numpy/core/src/multiarray/item_selection.c	(working tree)
@@ -28,7 +28,7 @@
 #include "alloc.h"
 #include "arraytypes.h"
 #include "array_coercion.h"
-#include "simd/simd.h"
+#include "contrib/python/numpy/py3/numpy/core/src/common/simd/simd.h"
 
 static NPY_GCC_OPT_3 NPY_INLINE int
 npy_fasttake_impl(
--- contrib/python/numpy/py3/numpy/core/src/umath/loops_arithm_fp.dispatch.c	(index)
+++ contrib/python/numpy/py3/numpy/core/src/umath/loops_arithm_fp.dispatch.c	(working tree)
@@ -16,1 +16,1 @@
-#include "simd/simd.h"
+#include "contrib/python/numpy/py3/numpy/core/src/common/simd/simd.h"
--- contrib/python/numpy/py3/numpy/core/src/umath/loops_arithmetic.dispatch.c	(index)
+++ contrib/python/numpy/py3/numpy/core/src/umath/loops_arithmetic.dispatch.c	(working tree)
@@ -18,1 +18,1 @@
-#include "simd/simd.h"
+#include "contrib/python/numpy/py3/numpy/core/src/common/simd/simd.h"
--- contrib/python/numpy/py3/numpy/core/src/umath/loops_exponent_log.dispatch.c	(index)
+++ contrib/python/numpy/py3/numpy/core/src/umath/loops_exponent_log.dispatch.c	(working tree)
@@ -20,1 +20,1 @@
-#include "simd/simd.h"
+#include "contrib/python/numpy/py3/numpy/core/src/common/simd/simd.h"
--- contrib/python/numpy/py3/numpy/core/src/umath/loops_trigonometric.dispatch.c	(index)
+++ contrib/python/numpy/py3/numpy/core/src/umath/loops_trigonometric.dispatch.c	(working tree)
@@ -15,1 +15,1 @@
-#include "simd/simd.h"
+#include "contrib/python/numpy/py3/numpy/core/src/common/simd/simd.h"
--- contrib/python/numpy/py3/numpy/core/src/umath/loops_unary_fp.dispatch.c	(index)
+++ contrib/python/numpy/py3/numpy/core/src/umath/loops_unary_fp.dispatch.c	(working tree)
@@ -22,1 +22,1 @@
-#include "simd/simd.h"
+#include "contrib/python/numpy/py3/numpy/core/src/common/simd/simd.h"
--- contrib/python/numpy/py3/numpy/core/src/_simd/_simd_inc.h	(index)
+++ contrib/python/numpy/py3/numpy/core/src/_simd/_simd_inc.h	(working tree)
@@ -12,1 +12,1 @@
-#include "simd/simd.h"
+#include "contrib/python/numpy/py3/numpy/core/src/common/simd/simd.h"
--- contrib/python/numpy/py3/numpy/core/src/umath/loops_umath_fp.dispatch.c	(index)
+++ contrib/python/numpy/py3/numpy/core/src/umath/loops_umath_fp.dispatch.c	(working tree)
@@ -12,1 +12,1 @@
-#include "simd/simd.h"
+#include "contrib/python/numpy/py3/numpy/core/src/common/simd/simd.h"
