--- contrib/python/numpy/py3/numpy/doc/__init__.py	(index)
+++ contrib/python/numpy/py3/numpy/doc/__init__.py	(working tree)
@@ -2,8 +2,10 @@ import os
 
 ref_dir = os.path.join(os.path.dirname(__file__))
 
-__all__ = sorted(f[:-3] for f in os.listdir(ref_dir) if f.endswith('.py') and
-           not f.startswith('__'))
+__all__ = [
+    "constants",
+    "ufuncs",
+]
 
 for f in __all__:
     __import__(__name__ + '.' + f)
