--- contrib/python/numpy/py3/numpy/core/src/umath/_rational_tests.c	(index)
+++ contrib/python/numpy/py3/numpy/core/src/umath/_rational_tests.c	(working tree)
@@ -1106,8 +1106,8 @@ rational_ufunc_test_add_rationals(char** args, npy_intp const *dimensions,
     }
 }
 
-
-PyMethodDef module_methods[] = {
+// XXX Arcadia confilct with contrib/python/pygit2
+PyMethodDef rational_module_methods[] = {
     {0} /* sentinel */
 };
 
@@ -1116,7 +1116,7 @@ static struct PyModuleDef moduledef = {
     "_rational_tests",
     NULL,
     -1,
-    module_methods,
+    rational_module_methods,
     NULL,
     NULL,
     NULL,
