--- contrib/python/numpy/py3/numpy/core/src/npymath/npy_math_private.h	(index)
+++ contrib/python/numpy/py3/numpy/core/src/npymath/npy_math_private.h	(working tree)
@@ -34,6 +34,10 @@
 #include "numpy/npy_endian.h"
 #include "numpy/npy_common.h"
 
+#ifdef _MSC_VER  // windows in ucrt\math.h makes #define complex _complex
+#undef complex
+#endif
+
 /*
  * The original fdlibm code used statements like:
  *      n0 = ((*(int*)&one)>>29)^1;             * index of high word *
--- contrib/python/numpy/py3/numpy/distutils/include/npy_cpu_dispatch_config.h	(index)
+++ contrib/python/numpy/py3/numpy/distutils/include/npy_cpu_dispatch_config.h	(working tree)
@@ -53,7 +53,11 @@
 #ifdef NPY__CPU_TARGET_POPCNT
 	/** POPCNT **/
 	#define NPY_HAVE_POPCNT 1
+#ifdef _MSC_VER
+	#include <nmmintrin.h>
+#else
 	#include <popcntintrin.h>
+#endif
 #endif /*NPY__CPU_TARGET_POPCNT*/
 #ifdef NPY__CPU_TARGET_SSE42
 	/** SSE42 **/
