--- contrib/python/numpy/py3/numpy/__init__.py	(index)
+++ contrib/python/numpy/py3/numpy/__init__.py	(working tree)
@@ -365,7 +365,7 @@ else:
     import sys
     if sys.platform == "darwin":
         with warnings.catch_warnings(record=True) as w:
-            _mac_os_check()
+            #_mac_os_check()
             # Throw runtime error, if the test failed Check for warning and error_message
             error_message = ""
             if len(w) > 0:
