--- contrib/python/numpy/py3/numpy/core/include/numpy/__multiarray_api.h	(index)
+++ contrib/python/numpy/py3/numpy/core/include/numpy/__multiarray_api.h	(working tree)
@@ -1,6 +1,9 @@
 
 #if defined(_MULTIARRAYMODULE) || defined(WITH_CPYCHECKER_STEALS_REFERENCE_TO_ARG_ATTRIBUTE)
 
+// XXX: dummy import to simplify static build
+static int import_array() { return 0; }
+
 typedef struct {
         PyObject_HEAD
         npy_bool obval;
