https://github.com/numpy/numpy/issues/20536
--- contrib/python/numpy/py3/numpy/core/src/multiarray/nditer_api.c	(index)
+++ contrib/python/numpy/py3/numpy/core/src/multiarray/nditer_api.c	(working tree)
@@ -1332,6 +1332,11 @@ NpyIter_GetInnerFixedStrideArray(NpyIter *iter, npy_intp *out_strides)
 
         for (iop = 0; iop < nop; ++iop) {
             stride = strides[iop];
+#if defined(__has_feature)
+#  if __has_feature(memory_sanitizer)
+            __msan_unpoison(&stride, sizeof(stride));
+#  endif
+#endif
             /*
              * Operands which are always/never buffered have fixed strides,
              * and everything has fixed strides when ndim is 0 or 1
