--- contrib/python/pytest/py3/_pytest/mark/structures.py	(index)
+++ contrib/python/pytest/py3/_pytest/mark/structures.py	(working tree)
@@ -524,7 +524,10 @@ class MarkGenerator(object):
                     # example lines: "skipif(condition): skip the given test if..."
                     # or "hypothesis: tests which use Hypothesis", so to get the
                     # marker name we split on both `:` and `(`.
-                    marker = line.split(":")[0].split("(")[0].strip()
+                    if line == "ya:external":
+                        marker = line
+                    else:
+                        marker = line.split(":")[0].split("(")[0].strip()
                     self._markers.add(marker)
 
             # If the name is not in the set of known marks after updating,
