commit fdd47fb34719d754d2ec6e0185947a9a1d880df8 (HEAD -> pytest-backport-pr-6607)
author: danila-eremin
date: 2022-01-22T16:55:31+03:00

    backport py3

--- contrib/python/pytest/py3/_pytest/python.py	(46b46cea857997531a7b9b4de32556be50bd5aea)
+++ contrib/python/pytest/py3/_pytest/python.py	(fdd47fb34719d754d2ec6e0185947a9a1d880df8)
@@ -896,7 +896,7 @@ class CallSpec2(object):
 
     @property
     def id(self):
-        return "-".join(map(safe_str, filter(None, self._idlist)))
+        return "-".join(map(safe_str, self._idlist))
 
     def setmulti2(self, valtypes, argnames, valset, id, marks, scopenum, param_index):
         for arg, val in zip(argnames, valset):
